### PR TITLE
fix(android/engine): Adjust default OSK landscape size

### PR DESCRIPTION
**Pull request by darcywong00**
Addresses the following point of #3225 
> Note, some other users request the OSK to be smaller (I assume landscape orientation?)
(4th point on https://community.software.sil.org/t/better-keyman-for-andriod/4791)

This PR updates the default OSK landscape sizes so it takes less of the screen.

This does not cover the spec for the TODO menu we plan to add in 15.0 for configuring the OSK height. Though this PR does add some hooks for updating the keyboard heights.

## Screenshots of landscape keyboards
Configurations show
* sil_euro_latin with suggestion banner
* khmer_angkor (4 rows, no banner)
* basic french (5 rows, no banner)

### Default Size (Nexus One emulator)
nexus one euro

nexus one khmer

nexus one french

### Pixel 4 emulator
pixel 4  euro

pixel 4 khmer

pixel 4 french

### 600dp (~7" device, Nexus 7 emulator)

nexus 7 euro

nexus 7 khmer

nexus 7 french

### 720dp (~10" device, Nexus 10 emulator)
nexus 10 euro

nexus 10 khmer

nexus 10 french

## User Testing
@MakaraSok - can you try the test build on your devices and comment if the keyboard sizes are too small, just right, or too big?

### CODE REVIEW

File: android/KMEA/app/src/main/assets/keyboard.html [hunk 1]
```diff
@@ -64,39 +64,41 @@
     // Update the KMW banner height
     function setBannerHeight(h) {
       var kmw=com.keyman.singleton;
+      var osk = kmw.osk;
       if (h > 0) {
-        var osk = kmw.osk;
-        osk.banner.height = Math.ceil(h / window.devicePixelRatio);
+        // The banner itself may not be loaded yet.  This will preemptively help set
+        // its eventual display height.
+        com.keyman.osk.Banner.DEFAULT_HEIGHT = h;
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/assets/keyboard.html:71]: @jahorton is this safe?

**Review by mcdurdin — COMMENTED** [review]
Looks reasonable, waiting for @MakaraSok to test and give feedback as well.

I note some logical changes in KMKeyboard.java and keyboard.html; can you explain why you are changing these?

File: android/KMEA/app/src/main/assets/keyboard.html (hunk 1 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/assets/keyboard.html:71]: I got this from iOS:

https://github.com/keymanapp/keyman/blob/8843828879fedb3a9f6302aa115a8b08e7994982/ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html#L79-L87

File: android/KMEA/app/src/main/assets/keyboard.html [hunk 2]
```diff
@@ -64,39 +64,41 @@
...
-      kmw['correctOSKTextSize']();
+      kmw.osk.show(true);
+      kmw.correctOSKTextSize();
     }
 
     function setOskWidth(w) {
-      if(w > 0) {
-        oskWidth = w;
-      }
+      oskWidth = w;
```
Comment (darcywong00) [inline android/KMEA/app/src/main/assets/keyboard.html:92]: Also from iOS
https://github.com/keymanapp/keyman/blob/8843828879fedb3a9f6302aa115a8b08e7994982/ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html#L99-L107

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java [hunk 3]
```diff
@@ -321,9 +321,10 @@ public void onConfigurationChanged(Configuration newConfig) {
     dismissKeyPreview(0);
     dismissSubKeysWindow();
     int bannerHeight = KMManager.getBannerHeight(context);
+    int oskHeight = KMManager.getKeyboardHeight(context);
     loadJavascript(KMString.format("setBannerHeight(%d)", bannerHeight));
     loadJavascript(KMString.format("setOskWidth(%d)", newConfig.screenWidthDp));
-    loadJavascript("setOskHeight(0)");
+    loadJavascript(KMString.format("setOskHeight(%d)", oskHeight));
```
Comment (darcywong00) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java:327]: This is groundwork for later when we dynamically adjust the OSK height

File: android/KMEA/app/src/main/assets/keyboard.html (hunk 2 above)
Comment (mcdurdin) [reply android/KMEA/app/src/main/assets/keyboard.html:92]: Although that snippet from iOS is for setting height, whereas this is width...?

Comment (darcywong00) [reply android/KMEA/app/src/main/assets/keyboard.html:92]: Oops.  here's setOskWidth

https://github.com/keymanapp/keyman/blob/f29c71d988e958b6dacb10f7c9be08b642864799/ios/engine/KMEI/KeymanEngine/resources/Keyman.bundle/Contents/Resources/keyboard.html#L109-L111

File: android/KMEA/app/src/main/assets/keyboard.html (hunk 1 above)
Comment (jahorton) [reply android/KMEA/app/src/main/assets/keyboard.html:71]: I don't think iOS has to worry about the `h / window.devicePixelRatio` part... are we sure that the pixel-ratio bit doesn't apply here?

**Comment by MakaraSok** [comment]
### Tested with 15.0.59 off of this build:
----

### Nexus One API 22, Android 5.1.1
* In app - everything is OK.
Screenshot_1622538499
Screenshot_1622538572
Screenshot_1622538543

* System keyboard - when switching from vertical to horizontal, the suggestion banner seem to be tilted up a bit out of alignment on EuroLatin keyboard. The ones without the suggestion banner are OK.
Screenshot_1622538798
Screenshot_1622538824
Screenshot_1622538771


### Pixel 4 API 21, Android 5.0.2
* In app - there is a black bar above the top row of the keyboard when selecting the keyboard on the vertical then switch to horizonal. This does not happen when switching the keyboard while in horizontal mode.
Screenshot_1622537437
Screenshot_1622537406
Screenshot_1622537421

* System keyboard - Everything is OK.
Screenshot_1622538186
Screenshot_1622538202
Screenshot_1622538193

### Nexus 7 API 22, Android 5.1.1
* In app - black bar above the keyboard when switching from vertical to horizontal
Screenshot_1622539571
Screenshot_1622539607
Screenshot_1622539591


* System keyboard - everything is OK.
Screenshot_1622539699
Screenshot_1622539745
Screenshot_1622539765


### Samsung Galaxy S9+, Android 10, One UI 2.5
* In app - there is a black space on the top row of each keyboard -- it looks like an empty banner
Screenshot_20210601-150527_Keyman
Screenshot_20210601-150351_Keyman
Screenshot_20210601-150539_Keyman

* System keyboard - Everything is OK.
Screenshot_20210601-153833_Messages
Screenshot_20210601-153905_Messages
Screenshot_20210601-153846_Messages

**Comment by darcywong00** [comment]
@MakaraSok - Can you use the globe button menu and exit? That seems to enforce KeymanWeb to refresh with the correct dimensions.

**Comment by mcdurdin** [comment]
So it sounds like from @MakaraSok's testing that some of the resizing is not working 100% yet -- is this behaviour new to this PR, Makara?

**Comment by MakaraSok** [comment]
> @MakaraSok - Can you use the globe button menu and exit? That seems to enforce KeymanWeb to refresh with the correct dimensions.

It was not a part of the test. Just it just now on my device, it does refresh the layout after closing it and get back to the keyboard.

Screenshot_20210602-092200_Keyman


But the predictive text seem to crash when flipping back vertical (see the screenshot below).

Screenshot1

Now flip it to horizontal to also see blank banner.

Screenshot2


> So it sounds like from @MakaraSok's testing that some of the resizing is not working 100% yet -- is this behaviour new to this PR, Makara?

I've never noticed the blank/void space above the keyboard before.

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Based on @MakaraSok's comments, the changes in keyboard.html appear to have changed behaviour for rotations?

**Comment by darcywong00** [comment]
> Based on @MakaraSok's comments, the changes in keyboard.html appear to have changed behaviour for rotations?

I've reverted keyboard.html (minor formatting difference now). @MakaraSok can you spot-check by retesting? Just 1-2 devices should be enough. I still see the banner clipped on rotation, so that's existing behavior.

**Comment by MakaraSok** [comment]
The new build `kMAPro-debug (15.0.61-alphatest5201)` has been tested on an emulators (a phone and tablet) and a physical device.

### Emulator - Pixel 2 - Android 10 - API 29

When loading the EuroLatin keyboard while on vertical view, everything is OK; but when tilt it to horizontal view the black bar above the keyboard appears:
Screenshot_1622703185

While staying in the horizontal view, press the globe key in an intend to have the keyboard layout refreshed (i.e. press on the globe key and then press on the keyboard name), and the black bar disappears returning the keyboard to the normal/expected state.
Screenshot_1622703196

Now, title the phone to vertical view, the banner is cropped off.
Screenshot_1622703205

Press the globe key to refresh and see the keyboard layout returns to normal again.
Screenshot_1622703218

The same behavior is observed when testing with French Basic keyboard and Khmer Angkor keyboard. But one weird thing is that when switching from the horizonal to the vertical view, the top row of the keyboard disappears.
Screenshot_1622704411
Screenshot_1622704034


Press the globe key to trigger the refresh does help to turn the keyboard to the normal/expected state. 
The top row of the keyboard stays disappear however if not pressing on the keyboard name after pressing the globe to refresh the layout. 

To see repro it:
1. load the French Basic keyboard while in vertical view and make sure that everything looks OK
2. title the keyboard to horizontal view to see the black bar on top of the keyboard.
3.1 press the globe key and press on the keyboard name to refresh 
3.2 the keyboard turn to normal, now title it to the vertical view to see the top cropped off
4.1 press the globe key and press else where other than the keyboard name
4.2 the black bar above the keyboard stays, now title it to the vertical view to see the keyboard looks OK, no top row cropped off.

### On physical device, Samsung Galaxy S9+, Android 10.
The steps 3.x and 4.x both refresh the keyboard and turn it to the normal state. As long as the globe key is pressed, no matter where you tap next, the keyboard will be refreshed and turned to be normal.

File: android/KMEA/app/src/main/assets/keyboard.html [hunk 4]
```diff
@@ -76,15 +76,15 @@
       if(h > 0) {
         oskHeight = h;
       }
-      var kmw=window['keyman'];
+      var kmw=com.keyman.singleton;
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/assets/keyboard.html:79]: Can we avoid using `com.keyman` in new code?

**Review by mcdurdin — CHANGES_REQUESTED** [review]

File: android/KMEA/app/src/main/assets/keyboard.html (hunk 4 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/assets/keyboard.html]: I ended up reverting the entire keyboard.html file

**Review by mcdurdin — APPROVED** [review]
Nice and tight now. LGTM. Rotation looks like it might be a priority to resolve.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.63-alpha](https://keyman.com/downloads/releases/alpha/15.0.63)

**Pull request merged**